core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java

```diff
@@ -5,49 +5,134 @@
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
+import org.eclipse.microprofile.config.ConfigProvider;
 import org.jboss.logging.Logger;
 
+import io.smallrye.config.SmallRyeConfig;
+
 public final class ContainerRuntimeUtil {
 
     private static final Logger log = Logger.getLogger(ContainerRuntimeUtil.class);
+    private static final String DOCKER_EXECUTABLE = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class)
+            .getOptionalValue("quarkus.native.container-runtime", String.class).orElse(null);
+    /*
+     * Caching the value in a file helps us only as much as one JVM execution is concerned.
+     * Test suite's pom.xml sets things like <argLine>-Djava.io.tmpdir="${project.build.directory}"</argLine>,
+     * so the file could appear in /tmp/ or C:\Users\karm\AppData\Local\Temp\ or in fact in
+     * quarkus/integration-tests/something/target.
+     * There is no point in reaching it in `Path.of(Paths.get("").toAbsolutePath().toString(), "target",
+     * "quarkus_container_runtime.txt")`
+     * as the file is deleted between JVM executions anyway.
+     */
+    static final Path CONTAINER_RUNTIME = Path.of(System.getProperty("java.io.tmpdir"), "quarkus_container_runtime.txt");
```

Comment:
why is it it need a file and not just a static var?

Reply:
IIUC, the problem is that we are using different CLs along the build so you end up loading things several time.

That being said, I'm also concerned we are using a file.
First, it won't work well in concurrent build environments as you are using a common file for all builds. That is a bad idea. You need to use `Files.createTempFile()`.
Second, my recollection is that using `deleteOnExit()` is not recommended. So you will need to figure out a better way of deleting the file at the end of the build.

Maybe we need to take a step back here? I'm not sure where this is used but could we use a build item to propagate the value? IIRC the build step will only be executed if it is consumed? Not sure it would be handy though.

Reply:
FWIW, I'm not comfortable releasing Alpha6 with this build concurrency issue in so I'm leaning towards reverting until we come up with a better solution.

So if you feel strongly that I shouldn't, please speak up!

FWIW, the version without the file seemed like a good transition version until we figure things out.

Reply:
I don't think the build item will work, because IIRC this is needed in places outside the build steps as well.

That said, using Files.createTempFile() does seem more reasonable

Reply:
Maybe we could just put it into a `static` field of a class that we always load from the system classloader?

FTR, one such field that is always loaded from the system CL is system properties :laughing:

/me hides

Reply:
So... thinking a bit more about it, using `Files.createTempFile()` in a static field won't work as you will still have different files for different CLs.

But @Ladicek what would be so wrong using a system property? It's a bit of a hack, sure, but it does actually sound more appealing than the other solutions. Would there be any obvious drawback?

Reply:
Hack is definitely is... But I think it's probable acceptable. 

On the other hand we could create a temp file based on each JVM run that would not be global to the whole system as the current on is

Reply:
@gsmet 
> IIUC, the problem is that we are using different CLs along
> the build so you end up loading things several time.

Yes, exactly. The situation had several stages:
- the current state is really suboptimal, many calls to `-v`, `--version`, `info` to get rootless state etc.
- then I got rid of perpetual `docker info` by having the rootlessness checked lazily only when really needed
- then I got rid of the superfluous `-v` calls, getting the docker command calls to even smaller number
- the last step is caching it in a file. The difference between this and using a static attribute is 1 superfluous docker version call

All in all, if I change the usage of a temporary file to a static variable, the PR would still be very beneficial in this aspect.

> First, it won't work well in concurrent build environments
> as you are using a common file for all builds.

Do you mean there would be a race with `StandardOpenOption.TRUNCATE_EXISTING`? If a build gets an empty file, it merely ignores it and defaults to calling docker again...

> You need to use Files.createTempFile()

I don't think so. That is obviously how temp files are used, but in this context I would end up where I started: Needing to remember somewhere the path to that file. I need to know the path. Not having to remember somewhere `/tmp/container_runtime_<some arbitrary generated number I don't know>.txt`. So that is why I did what I did with the file name.

> recollection is that using deleteOnExit() is not recommended

Could you elaborate? The only problem with deleteOnExit is that is is not guaranteed and it is platform specific, so it's a very bad idea to have it if you need to rely on the file *being really gone*. This is not the case. The worst that might happen (IMHO) is that the file stays there in temp, user switches from docker to podman and there would be a warning or failure depending on what user does. Perhaps we could time stamp it it mention it in the log...

> Maybe we need to take a step back here?
> I'm not sure where this is used but could we use a build item
> to propagate the value?

I am all for taking a step back. I spent some nontrivial time on Windows and Linux, considering dev mode, docker-image extension usage, native build using builder image and test suite using runtime containers. Dumping the info into a file was the thing that looked the least complicated from my limited POV. I am happy to try other avenues.

Reply:
yeah so build concurency was my concern too about the filename.

Though I'm not sure its that big a deal in practice as what will happen is a concurrent one will pick up the one running on the same system. Only time it will problematic is if those two runs are running with different default environment where one would have picked podman over docker and vice versa. 

One way to avoid that issue is to use `ProcessHandle.current().pid()` and append/prefix the filename. Then at least it will be per process rather than global.

About deleteOnExit() - my understanding is that you just shouldn't assume it always gets deleted (i.e. ctrl+c will hard stop and deletes does not happen); but no other mechanism can do that better anyway.

Reply:
@Karm for now I haven't seen any strong opinions against a system property approach? So if it works, maybe we could try that? That would alleviate all the concerns expressed so far regarding the file approach?

Maybe shoe it in with your current code to see if it works before cleaning things up.

Reply:
Something like `quarkus-local-container-runtime` (lowercase, no `quarkus.` prefix) to make sure it doesn't collide with the config properties.